### PR TITLE
DL: Improve parameter parsing and add unit tests

### PR DESCRIPTION
**Pull request by orhankislal**
JIRA: MADLIB-1336

This commit ensures that metrics parameter can be skipped by the user.
It also enforces that optimizer and loss are required.

This commit also fixes the fit params parsing to allow NULL (no user defined
fit params)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/875/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/877/

**Comment by orhankislal** [comment]
The failing unit-test will be fixed by the https://github.com/apache/madlib/pull/383

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/878/

File: src/ports/postgres/modules/deep_learning/test/unit_tests/test_madlib_keras.py_in [hunk 1]
```diff
@@ -423,13 +423,23 @@ class MadlibKerasFitTestCase(unittest.TestCase):
         opt_name,opt_args,result_params = self.subject.parse_and_validate_compile_params(test_str)
         self.assertDictEqual(result_params, compile_dict)
 
+        test_str = "optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss='categorical_crossentropy'"
+        compile_dict = {'optimizer':'SGD(lr=0.01, decay=1e-6, nesterov=True)', 'loss':'categorical_crossentropy'}
+        opt_name,opt_args,result_params = self.subject.parse_and_validate_compile_params(test_str)
+        self.assertDictEqual(result_params, compile_dict)
+
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/test/unit_tests/test_madlib_keras.py_in]: Noticed we don't have any unit tests for `convert_string_of_args_to_dict()`. If you think it's out of this PR's scope, can you kindly create a JIRA to keep track of it, or add test cases as part of this PR if not?

**Comment by kaknikhil** [comment]
@njayaram2 
PR is to be re reviewed. We add a few more unit tests that will cover the test cases for `convert_string_of_args_to_dict`

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/887/

File: src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in [hunk 2]
```diff
@@ -318,7 +318,7 @@ SELECT madlib_keras_fit(
     'independent_var',
     'model_arch',
     1,
-    $$ optimizer=Adam(epsilon=None), loss=losses.categorical_crossentropy, metrics=['accuracy']$$::text,
+    $$ optimizer=Adam(epsilon=None), loss='categorical_crossentropy', metrics=['accuracy']$$::text,
```
[the remark this answers is not in the document]
Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in]: we do not support the losses prefix (or any other prefix). The user has to pass in a string with the name of the unqualified loss function .

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/889/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/890/

**Comment by fmcquillan99** [comment]
I guess I am not clear on what are mandatory params and what are optional params, and how errors are handled.



(1)
regular test
```
DROP TABLE IF EXISTS model1, model1_summary;
SELECT madlib.madlib_keras_fit('mnist_train_packed',  -- source_table
                               'model1', -- model
                               'dependent_var',
                               'independent_var',
                               'model_arch_library',  -- model_arch_table
                               1,  -- model_arch_id
                               $$ loss='categorical_crossentropy', optimizer='adam', metrics=['acc'] $$,  -- compile_params
                               $$ batch_size=128, epochs=4 $$,  -- fit_params
                               5,  -- num_iterations
                               0,  -- gpus per host
                               NULL,  -- validation_table
                               'Frank',  -- name
                               'A test model'  -- description
                              );

INFO:  Processed 30000 images: Fit took 4.37503194809 sec, Total was 5.91591286659 sec  (seg0 slice1 10.128.0.41:40000 pid=2400)
CONTEXT:  PL/Python function "fit_transition"
INFO:  Processed 30000 images: Fit took 4.39241290092 sec, Total was 5.95022010803 sec  (seg1 slice1 10.128.0.41:40001 pid=2401)
CONTEXT:  PL/Python function "fit_transition"
INFO:  Time for iteration 5: 8.3350250721 sec
CONTEXT:  PL/Python function "madlib_keras_fit"
INFO:  Average loss after training iteration 5: 0.0441886410117
CONTEXT:  PL/Python function "madlib_keras_fit"
INFO:  Average accuracy after training iteration 5: 0.986333310604
```
OK

(2)
no metrics param
```
DROP TABLE IF EXISTS model1, model1_summary;
SELECT madlib.madlib_keras_fit('mnist_train_packed',  -- source_table
                               'model1', -- model
                               'dependent_var',  -- dependent_varname
                               'independent_var',  -- independent_varname
                               'model_arch_library',  -- model_arch_table
                               1,  -- model_arch_id
                               $$ loss='categorical_crossentropy', optimizer='adam' $$,  -- compile_params
                               $$ batch_size=128, epochs=4 $$,  -- fit_params
                               1,  -- num_iterations
                               0,  -- gpus per host
                               NULL,  -- validation_table
                               'Frank',  -- name
                               'A test model'  -- description
                              );

INFO:  Model architecture size: 1KB
CONTEXT:  PL/Python function "madlib_keras_fit"
INFO:  Model state (serialized) size: 0MB
CONTEXT:  PL/Python function "madlib_keras_fit"
ERROR:  plpy.SPIError: KeyError: 'acc' (plpython.c:5038)  (seg1 slice1 10.128.0.41:40001 pid=3000) (plpython.c:5038)
DETAIL:  
Traceback (most recent call last):
  PL/Python function "fit_transition", line 6, in <module>
    return madlib_keras.fit_transition(**globals())
  PL/Python function "fit_transition", line 431, in fit_transition
PL/Python function "fit_transition"
CONTEXT:  Traceback (most recent call last):
  PL/Python function "madlib_keras_fit", line 21, in <module>
    madlib_keras.fit(**globals())
  PL/Python function "madlib_keras_fit", line 173, in fit
PL/Python function "madlib_keras_fit"                          
```
Is this the right error?


(3)
no loss param

DROP TABLE IF EXISTS model1, model1_summary;
SELECT madlib.madlib_keras_fit('mnist_train_packed',  -- source_table
                               'model1', -- model
                               'dependent_var',
                               'independent_var',
                               'model_arch_library',  -- model_arch_table
                               1,  -- model_arch_id
                               $$ optimizer='adam', metrics=['acc'] $$,  -- compile_params
                               $$ batch_size=128, epochs=4 $$,  -- fit_params
                               5,  -- num_iterations
                               0,  -- gpus per host
                               NULL,  -- validation_table
                               'Frank',  -- name
                               'A test model'  -- description
                              );

INFO:  Model architecture size: 1KB
CONTEXT:  PL/Python function "madlib_keras_fit"
INFO:  Model state (serialized) size: 0MB
CONTEXT:  PL/Python function "madlib_keras_fit"
ERROR:  plpy.SPIError: plpy.Error: loss is a required parameter for compile (plpython.c:5038)  (seg0 slice1 10.128.0.41:40000 pid=3238) (plpython.c:5038)
DETAIL:  
Traceback (most recent call last):
  PL/Python function "fit_transition", line 6, in <module>
    return madlib_keras.fit_transition(**globals())
  PL/Python function "fit_transition", line 408, in fit_transition
  PL/Python function "fit_transition", line 101, in compile_and_set_weights
  PL/Python function "fit_transition", line 274, in compile_model
  PL/Python function "fit_transition", line 175, in parse_and_validate_compile_params
  PL/Python function "fit_transition", line 96, in _assert
PL/Python function "fit_transition"
CONTEXT:  Traceback (most recent call last):
  PL/Python function "madlib_keras_fit", line 21, in <module>
    madlib_keras.fit(**globals())
  PL/Python function "madlib_keras_fit", line 173, in fit
PL/Python function "madlib_keras_fit"
```
Is this the right error to throw

(4)
No fit params

```
DROP TABLE IF EXISTS model1, model1_summary;
SELECT madlib.madlib_keras_fit('mnist_train_packed',  -- source_table
                               'model1', -- model
                               'dependent_var',
                               'independent_var',
                               'model_arch_library',  -- model_arch_table
                               1,  -- model_arch_id
                               $$ loss='categorical_crossentropy', optimizer='adam', metrics=['acc'] $$,  -- compile_params
                               NULL,  -- fit_params
                               5,  -- num_iterations
                               0,  -- gpus per host
                               NULL,  -- validation_table
                               'Frank',  -- name
                               'A test model'  -- description
                              );

ERROR:  TypeError: cannot concatenate 'str' and 'NoneType' objects (plpython.c:5038)
CONTEXT:  Traceback (most recent call last):
  PL/Python function "madlib_keras_fit", line 21, in <module>
    madlib_keras.fit(**globals())
  PL/Python function "madlib_keras_fit", line 142, in fit
PL/Python function "madlib_keras_fit"
```
I would have thought that defaults would be used.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/898/

**Comment by orhankislal** [comment]
@fmcquillan99 Thanks for the comments. The scope of this PR has changed a bit. I changed the title to reflect it. 
1. no metrics: The parsing code allows dropping the metrics parameter but a different part of the code still requires it. A future PR will drop that requirement which will fix the issue. Currently, we do require a metric.
1. no loss: The error message is correct.
1. no fit: There was a minor bug which should be fixed now. We will allow both NULL and an empty string for fit_params.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/910/

**Comment by fmcquillan99** [comment]
LGTM

**Pull request closed without merging**